src/mono/mono/mini/method-to-ir.c

```diff
@@ -7065,6 +7065,63 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 				addr = mono_emit_jit_icall (cfg, mono_get_native_calli_wrapper, args);
 			}
 
+			if (!method->dynamic && fsig->pinvoke &&
+			    !method->wrapper_type) {
+				/* MONO_WRAPPER_DYNAMIC_METHOD dynamic method handled above in the
+				method->dynamic case; for other wrapper types assume the code knows
+				what its doing and added its own GC transitions */
+
+				/* TODO: unmanaged[SuppressGCTransition] call conv will set
+				 * skip_gc_trans to TRUE*/
+				gboolean skip_gc_trans = FALSE;
+				if (!skip_gc_trans) {
+#if 0
+					fprintf (stderr, "generating wrapper for calli in method %s with wrapper type %s\n", method->name, mono_wrapper_type_to_str (method->wrapper_type));
+#endif
+					/* Call the wrapper that will do the GC transition instead */
+					MonoMethod *wrapper = mono_marshal_get_native_func_wrapper_indirect (method->klass, fsig, cfg->compile_aot);
+					/* box the indirect function ptr */
+					/* FIXME: it would be nice if the wrapper didn't require a boxed argument */
+
```

Comment:
Why does this need to be boxed ?

Reply:
I'm not sure. I'm reusing the IL generation in `emit_native_wrapper_ilgen` - and when it has to do an indirect call it expects the function to be boxed for some reason (I think it's used for delegates and it's more uniform if the first arg is always boxed.  but I'm not sure).

I didn't want to change it in this PR until I saw green tests, but it probably wouldn't be too hard to change it

Reply:
This is because of the 'func_param' argument to emit_native_wrapper_ilgen which is used by delegate invokes. It probably needs another param which says that it receives an unboxed value.